commit 1d50099a1a160f5cf212fcce22c25276bbd912c5
Author: Jonathan Lebon <jlebon@redhat.com>
Date:   Fri May 30 11:47:51 2014 -0400

    BZ1099133: prologue searching: check previous line too
    
    We perform binary search to retrieve the line record matching the
    entrypc. However, we may by chance fall on the second of two line
    records for the same entrypc. Check for this after binary search is
    done.

diff --git a/dwflpp.cxx b/dwflpp.cxx
index 09407b4..ff3c1a6 100644
--- a/dwflpp.cxx
+++ b/dwflpp.cxx
@@ -2271,6 +2271,20 @@ dwflpp::resolve_prologue_endings (func_info_map_t & funcs)
             else if (addr < entrypc) { l = entrypc_srcline_idx; }
             else { h = entrypc_srcline_idx; }
           }
+
+        // We may by chance have fallen on the second of two consecutive line
+        // records for the same addr. If the previous Dwarf_Line is indeed at
+        // the same address, pick that one instead. It means that there is no
+        // prologue, which the code that follows will soon find. (BZ1099133).
+        if (entrypc_srcline && entrypc_srcline_idx > 0)
+          {
+            Dwarf_Line* lr = dwarf_onesrcline(lines, entrypc_srcline_idx-1);
+            if (DWARF_LINEADDR(lr) == entrypc)
+              {
+                entrypc_srcline = lr;
+                entrypc_srcline_idx--;
+              }
+          }
       }
       if (!entrypc_srcline)
         {
